lib/ansible/module_utils/common/parameters.py

```diff
@@ -724,14 +724,15 @@ def _validate_sub_spec(argument_spec, parameters, prefix='', options_context=Non
             options_context.append(param)
 
             # Make sure we can iterate over the elements
-            if isinstance(parameters[param], dict):
+            if not isinstance(parameters[param], list):
                 elements = [parameters[param]]
             else:
                 elements = parameters[param]
 
             for idx, sub_parameters in enumerate(elements):
                 if not isinstance(sub_parameters, dict):
-                    errors.append(SubParameterTypeError("value of '%s' must be of type dict or list of dicts" % param))
+                    errors.append(SubParameterTypeError("value of '%s' is a sub-parameter and must be of type dict or list of dicts" % param))
```

Comment:
Maybe the message should use the `wanted` var. I edited the error in attempt to make it clearer that suboptions need to be a dict or list of dicts, but if wanted == 'dict', then a list would still cause an error (and vice versa) for the specific suboption. Not sure if this clarifies that well.